swupd/packs.go

```diff
@@ -14,11 +14,306 @@
+		hdr.Linkname = link
+	case typeFile:
+		if !fi.Mode().IsRegular() {
+			return fmt.Errorf("couldn't use %s for packing: manifest expected a regular file but it is not", realname), true
+		}
+		hdr.Typeflag = tar.TypeReg
+	default:
+		return fmt.Errorf("unsupported file %s in chroot with type %q", f.Name, f.Type), true
+	}
+
+	// After we start writing on the tar writer, we can't let the caller fallback to another
+	// option anymore.
+
+	err = tw.WriteHeader(hdr)
+	if err != nil {
+		return err, false
+	}
+
+	if hdr.Typeflag == tar.TypeReg {
+		realfile, err := os.Open(realname)
+		if err != nil {
+			return err, false
+		}
+		defer realfile.Close()
+
+		_, err = io.Copy(tw, realfile)
+		if err != nil {
+			return err, false
+		}
+	}
+	return nil, false
+}
+
+type compressedTarReader struct {
+	*tar.Reader
+	compressionCloser io.Closer
+}
+
+func (ctr *compressedTarReader) Close() error {
+	if ctr.compressionCloser != nil {
+		return ctr.compressionCloser.Close()
+	}
+	return nil
+}
+
+// Compression algorithms have "magic" bytes in the beginning of the file to identify them.
+var (
+	gzipMagic  = []byte{0x1F, 0x8B}
+	xzMagic    = []byte{0xFD, '7', 'z', 'X', 'Z', 0x00}
+	bzip2Magic = []byte{'B', 'Z', 'h'}
+)
+
+// newCompressedTarReader creates a struct compatible with tar.Reader reading from uncompressed or
+// compressed input.
+func newCompressedTarReader(rs io.ReadSeeker) (*compressedTarReader, error) {
+	var h [6]byte
+	_, err := rs.Read(h[:])
+	if err != nil {
+		return nil, err
+	}
+	_, err = rs.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	result := &compressedTarReader{}
+
+	switch {
+	case bytes.HasPrefix(h[:], gzipMagic):
+		gr, err := gzip.NewReader(rs)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't decompress using gzip: %s", err)
+		}
+		result.compressionCloser = gr
+		result.Reader = tar.NewReader(gr)
+	case bytes.HasPrefix(h[:], xzMagic):
+		xr, err := newExternalReader(rs, "unxz")
+		if err != nil {
+			return nil, fmt.Errorf("couldn't decompress using xz: %s", err)
+		}
+		result.compressionCloser = xr
+		result.Reader = tar.NewReader(xr)
+	case bytes.HasPrefix(h[:], bzip2Magic):
+		br := bzip2.NewReader(rs)
+		result.Reader = tar.NewReader(br)
+	default:
+		// Assume uncompressed tar and let it complain if not valid.
+		result.Reader = tar.NewReader(rs)
+	}
+	return result, nil
+}
+
+func copyFromFullfile(tw *tar.Writer, outputDir string, f *File) error {
+	fullfilePath := filepath.Join(outputDir, fmt.Sprintf("%d", f.Version), "files", f.Hash.String()+".tar")
+	fullfile, err := os.Open(fullfilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open fullfile for %s in version %d: %s", f.Name, f.Version, err)
+	}
+	defer fullfile.Close()
+
+	fullfileReader, err := newCompressedTarReader(fullfile)
+	if err != nil {
+		return fmt.Errorf("failed to read fullfile %s: %s", fullfilePath, err)
+	}
+	// TODO: Check for error when closing this.
+	defer fullfileReader.Close()
```

Comment:
might as well check the error here the same way you do elsewhere.

Reply:
Done.